### DISCUSSION TITLE
Useing reaktoro for enthalpy and latent heat calculations

### DISCUSSION THREAD

**Discussion by avdudchenko** [reaktoro/reaktoro#408] [Q&A]
### Description

We have been trying to use Reaktoro for enthalpy calculations, specifically for saline waters (such as sea water) for applications such as thermal desalination. In these systems we are mostly interested in energy required to heat a solution from T1 to T2 while salinity changes. 

When using enthalpy calculations in Reaktoro where only temperature changes, we get expected behavior. 

But when we, for example, change feed salinity (through evaporation of water, desalination via membrane-based process), we find that enthalpy changes do not seem to follow expected trends as those shown by Sharqawy et al. for seawater, unless we use heat capacities instead. 

Minmal example code
```py

import reaktoro as rkt

water_spec = {
    "Na+": 10556,
    "K+": 380,
    "Ca+2": 400,
    "Mg+2": 1262,
    "Cl-": 17000,
    "SO4-2": 2649,
    "HCO3-": 140,
}
mw_spec = {
    "Na+": 22.989769e3,
    "Ca+2": 40.078e3,
    "SO4-2": 40.078e3,
    "Cl-": 35.453e3,
    "HCO3-": 61.0168e3,
    "K+": 39.0983e3,
    "Mg+2": 39.0983e3,
}

temperature = 50

db = rkt.SupcrtDatabase("supcrtbl")
species = " ".join(list(water_spec.keys()))
solution = rkt.AqueousPhase(rkt.speciate(species))
solution.set(rkt.ActivityModelPitzer())

system = rkt.ChemicalSystem(db, solution)
state_DI = rkt.ChemicalState(system)
state_SW = rkt.ChemicalState(system)
tds = 0

for ion, value in water_spec.items():
    state_DI.set(ion, 1e-8 / mw_spec[ion], "mol")
    state_SW.set(ion, value / mw_spec[ion], "mol")
    tds += value / 1000 / 1000  # needs to be in kg/kg
state_DI.set("H2O(aq)", 1000 / 18.01528, "mol")
state_SW.set("H2O(aq)", 1000 / 18.01528, "mol")
state_DI.temperature(temperature, "celsius")
state_SW.temperature(temperature, "celsius")
state_DI.pressure(1, "atm")
state_SW.pressure(1, "atm")

rkt.equilibrate(state_DI)
rkt.equilibrate(state_SW)

DI_enthalpy = state_DI.props().specificEnthalpy() / 1000
SW_enthalpy = state_SW.props().specificEnthalpy() / 1000
DI_cp = state_DI.props().specificHeatCapacityConstP() / 1000 * temperature
SW_cp = state_SW.props().specificHeatCapacityConstP() / 1000 * temperature

delta_sharqway = (
    -(tds * (27062.623 + tds) + tds * (4835.675 + tds) * temperature) / 1000
)
print(
    tds,
    "reaktoro via specific enthalpy:",
    SW_enthalpy-DI_enthalpy ,
    "kJ/kg",
    "reaktoro via specific heat capacity:",
    SW_cp - DI_cp,
    "Seawater Sharqawy:",
    delta_sharqway,
    "kJ/kg",
)

```
When using enthalpy directly, we get a difference between two solutions that is orders of magnitude off.

Using heat capacity produces much better agreement, so it seems we are not correctly referencing system here- or using correct approach. 

`reaktoro via specific enthalpy: 275.503 kJ/kg reaktoro via specific heat capacity: -9.00035 Seawater Sharqawy: -8.707180977157218 kJ/kg`

In same train of thought, we wanted to use Reaktoro to get at heat of vaporization, which one would assume could be retrieved by getting enthalpy of saline solution, and enthalpy of a vapor at same temperature. This does not produce expected result either, if using speciesStandardEnthalpy it's possible to get something closer in magnitude, but the values are still far apart (to tune of 200-300 kJ/kg, and only gets worse with increasing salinity). Minimal code below:

```py

import reaktoro as rkt
water_spec = {
    "Na+": 10556,
    "K+": 380,
    "Ca+2": 400,
    "Mg+2": 1262,
    "Cl-": 17000,
    "SO4-2": 2649,
    "HCO3-": 140,
}
mw_spec = {
    "Na+": 22.989769e3,
    "Ca+2": 40.078e3,
    "SO4-2": 40.078e3,
    "Cl-": 35.453e3,
    "HCO3-": 61.0168e3,
    "K+": 39.0983e3,
    "Mg+2": 39.0983e3,
}

temperature = 50

db = rkt.SupcrtDatabase("supcrtbl")
species = " ".join(list(water_spec.keys()))
solution = rkt.AqueousPhase(rkt.speciate(species))
solution.set(rkt.ActivityModelPitzer())

system = rkt.ChemicalSystem(db, solution)

gas_phase = rkt.GaseousPhase("H2O(g)")
gas_phase.setActivityModel(rkt.ActivityModelRedlichKwong())
gas_system = rkt.ChemicalSystem(db, gas_phase)
state_vapor = rkt.ChemicalState(gas_system)
state_SW = rkt.ChemicalState(system)
tds = 0

for ion, value in water_spec.items():
    state_SW.set(ion, value / mw_spec[ion], "mol")
    tds += value / 1000 / 1000  # needs to be in kg/kg
state_vapor.set("H2O(g)", 1000 / 18.01528, "mol")
state_SW.set("H2O(aq)", 1000 / 18.01528, "mol")
state_vapor.temperature(temperature, "celsius")
state_SW.temperature(temperature, "celsius")
state_vapor.pressure(1, "atm")
state_SW.pressure(1, "atm")

rkt.equilibrate(state_vapor)
rkt.equilibrate(state_SW)

Vapor_enthalpy = state_vapor.props().specificEnthalpy() / 1000
SW_enthalpy = state_SW.props().specificEnthalpy() / 1000
Enth_h2ov = (
    state_vapor.props().speciesStandardEnthalpy("H2O(g)")
    * state_vapor.props().speciesAmount("H2O(g)")
    / 1000
) / state_vapor.props().mass()

heat_of_vaporization = 2381.9 * (1 - tds)
print(
    tds,
    "reaktoro direct enthaply:",
    Vapor_enthalpy - SW_enthalpy,
    "reaktoro species enthalpy:",
    Enth_h2ov - SW_enthalpy,
    "heat of vaporization Sharqawy:",
    heat_of_vaporization,
)

```

This produces the following output:
`reaktoro direct enthaply: 47.6062 reaktoro species enthalpy: 2110.15 heat of vaporization Sharqawy: 2304.7574047000003`

I belive both of these issues are related and stem from our misunderstanding of how to properly use enthalpy values returned by Reaktoro. 

Any advice/suggestions would be greatly appreciated. 

Reference to seawater thermophysical properties: Sharqawy, M. H., Lienhard, J. H. & Zubair, S. M. Thermophysical properties of seawater: a review of existing correlations and data. Desalination and Water Treatment 16, 354–380 (2010).

### What is your operating system?

None

**Comment by allanleal** [comment]
Hi @avdudchenko - thanks for sharing this. Here are some thoughts:

* **Mixing SUPCRT thermodynamic database with Pitzer activity model with parameters from PHREEQC**: By default, `ActivityModelPitzer` will use the interaction parameters from the PHREEQC database `pitzer.dat`. Ideally, you would want to be consistent and use the same `pitzer.dat` database for computing the standard-state thermodynamic properties of the species (via logK of reactions). In this case, you would instead use a `PhreeqcDatabase("pitzer.dat")` object. However, to be consistent with PHREEQC, a simpler equation of state of water is used, whereas the use of SUPCRT databases in Reaktoro will use the more accurate water EOS of Wagner-Pruss (2002). 

* **Databases may not have been calibrated with enthalpy-related experimental data**: In this case, enthalpy calculations for concentrated electrolyte/brine solutions may not be accurate. 

*  **Mismatch reference states for enthalpy**: The SUPCRT databases rely on the HKF model to compute the standard-state thermodynamic properties of the species, which are ultimately used to compute the enthalpy of the phase (please check implementation in `Reaktoro/Core/ChemicalPropsPhase.hpp`). The HKF model assumes the triple point of water as the reference state. The issue you are experiencing is thus related to this discussion #406. 

I think the last item is the most decisive here to pinpoint the mismatch you are observing. 

One last thing: For your computation of enthalpy using heat capacity, I think this needs revision: 1) use delta of temperature, from 25 °C and 2) add the enthalpy at 25 °C. 

WDYT?

**Comment by allanleal** [comment]
Hi @avdudchenko - the script below introduces a correction to remove the contribution of the enthalpy of water at the triple point, introduced in the HKF model. See line of code below:

https://github.com/reaktoro/reaktoro/blob/eef365989737ae9f3626fd1041e7476839386680/Reaktoro/Models/StandardThermoModels/StandardThermoModelWaterHKF.cpp#L46

However, even after this correction, we obtain a delta enthalpy due to concentration changes of -273.647 kJ/kg whereas the model you are comparing to produces -8.7072 kJ/kg. Maybe worth checking at lower TDS. Perhaps, as I said, the SUPCRT database was not calibrated with same enthalpy data used to regress the model you and referring to (or I did not get the correction right!). 

```py
import reaktoro as rkt

water_spec = {
    "Na+": 10556,
    "K+": 380,
    "Ca+2": 400,
    "Mg+2": 1262,
    "Cl-": 17000,
    "SO4-2": 2649,
    "HCO3-": 140,
}
mw_spec = {
    "Na+": 22.989769e3,
    "Ca+2": 40.078e3,
    "SO4-2": 40.078e3,
    "Cl-": 35.453e3,
    "HCO3-": 61.0168e3,
    "K+": 39.0983e3,
    "Mg+2": 39.0983e3,
}

temperature = 50

db = rkt.SupcrtDatabase("supcrtbl")
species = " ".join(list(water_spec.keys()))
solution = rkt.AqueousPhase(rkt.speciate(species))
solution.set(rkt.ActivityModelPitzer())

system = rkt.ChemicalSystem(db, solution)
state_DI = rkt.ChemicalState(system)
state_SW = rkt.ChemicalState(system)
tds = 0

for ion, value in water_spec.items():
    state_DI.set(ion, 1e-8 / mw_spec[ion], "mol")
    state_SW.set(ion, value / mw_spec[ion], "mol")
    tds += value / 1000 / 1000  # needs to be in kg/kg
state_DI.set("H2O(aq)", 1000 / 18.01528, "mol")
state_SW.set("H2O(aq)", 1000 / 18.01528, "mol")
state_DI.temperature(temperature, "celsius")
state_SW.temperature(temperature, "celsius")
state_DI.pressure(1, "atm")
state_SW.pressure(1, "atm")

rkt.equilibrate(state_DI)
rkt.equilibrate(state_SW)

Mw = 0.01801528  # kg/mol

Hwtr_molar = -287.721128   # enthalpy of water at the triple point -- kJ/mol
Hwtr_kg = Hwtr_molar / Mw   # enthalpy of water at the triple point -- kJ/kg

H_SW_molar = state_SW.props().molarEnthalpy() / 1000  # in kJ/mol
H_DI_molar = state_DI.props().molarEnthalpy() / 1000  # in kJ/mol

xH2O_SW = state_SW.props().speciesMoleFraction("H2O(aq)")
xH2O_DI = state_DI.props().speciesMoleFraction("H2O(aq)")

H_SW_molar_corrected = H_SW_molar - xH2O_SW * Hwtr_molar  # in kJ/mol
H_DI_molar_corrected = H_DI_molar - xH2O_DI * Hwtr_molar  # in kJ/mol

H_SW_kg_corrected = H_SW_molar_corrected * state_SW.props().amount() / state_SW.props().mass()
H_DI_kg_corrected = H_DI_molar_corrected * state_DI.props().amount() / state_DI.props().mass()

delta_sharqway = (
    -(tds * (27062.623 + tds) + tds * (4835.675 + tds) * temperature) / 1000
)

print(
    tds,
    "reaktoro via specific enthalpy:",
    H_SW_kg_corrected-H_DI_kg_corrected ,
    "kJ/kg",
    "Seawater Sharqawy:",
    delta_sharqway,
    "kJ/kg",
)

# Output:
# 0.032387 reaktoro via specific enthalpy: -273.647 kJ/kg Seawater Sharqawy: -8.707180977157218 kJ/kg
```
